pennylane/data/data_manager/foldermap.py

```diff
@@ -0,0 +1,177 @@
+        """
+        try:
+            return self.__curr_level["__default"]
+        except KeyError as exc:
+            raise ValueError("No default available for parameter") from exc
+
+    def find(
+        self,
+        data_name: str,
+        missing_default: Optional[ParamArg] = ParamArg.DEFAULT,
+        **params: Union[typing.Iterable[ParamVal], ParamArg],
+    ) -> List[Tuple[Description, DataPath]]:
+        """Returns a 2-tuple of dataset description and paths, for each dataset that
+        matches ``params``."""
+
+        try:
+            data_names_to_params = self.__curr_level["__params"]
+        except KeyError as exc:
+            raise RuntimeError("Can only call find() from top level of foldermap") from exc
+
+        try:
+            param_names: List[str] = data_names_to_params[data_name]
+        except KeyError as exc:
+            raise ValueError(f"No datasets with data name: '{data_name}'") from exc
+
+        curr: List[Tuple[Description, Union[FolderMapView, DataPath]]] = [
+            (Description(()), self[data_name])
+        ]
+        todo: List[Tuple[Description, Union[FolderMapView, DataPath]]] = []
+        done: List[Tuple[Description, DataPath]] = []
+
+        for param_name in param_names:
+            param_arg = params.get(param_name, missing_default)
+            if param_arg is None:
+                raise ValueError(f"Missing argument for parameter '{param_name}'")
+
+            while curr:
+                curr_description, curr_level = curr.pop()
+                if isinstance(curr_level, DataPath):
+                    done.append((curr_description, curr_level))
+                    continue
+
+                if param_arg == ParamArg.FULL:
+                    next_params = curr_level
+                elif param_arg == ParamArg.DEFAULT:
+                    next_params = (curr_level.get_default_key(),)
+                elif isinstance(param_arg, str):
+                    next_params = (param_arg,)
+                else:
+                    next_params = param_arg
+
+                todo.extend(
+                    (
+                        Description((*curr_description.items(), (param_name, next_param))),
+                        curr_level[next_param],
+                    )
+                    for next_param in next_params
+                )
+
+            curr, todo = todo, curr
+
+        done.extend(curr)
+
+        return done
+
+    def __getitem__(
+        self, __key: Union[str, Literal[ParamArg.DEFAULT]]
+    ) -> Union["FolderMapView", DataPath]:
+        """Gets the item with key. If key is ``ParamArg.DEFAULT``, return the
+        item under the default parameter, or raise a ``ValueError`` if no
+        default exists."""
+        if __key in self.__PRIVATE_KEYS:
+            raise KeyError(__key)
+
+        if __key == ParamArg.DEFAULT:
+            return self[self.get_default_key()]
+
+        elem = self.__curr_level[__key]
```

Comment:
can we do more handling around this? in the old dataset, when you gave an invalid value for a key, you would be told what options are available. now you just get a keyerror. For example:

master:
```pycon
>>> qml.data.load("qchem", molname="foo", basis="STO-3G", bondlength=0.86)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/data_manager.py", line 253, in load
    _validate_params(data_name, description, attributes)
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/data_manager.py", line 116, in _validate_params
    raise exc  # pylint:disable=raising-bad-type
ValueError: molname value of 'foo' is not available. Available values are ['C2H4', 'H2', 'H3+', 'OH-', 'H6', 'C2', 'HeH+', 'CH4', 'BH3', 'H4', 'CO', 'HF', 'O2', 'He2', 'LiH', 'BeH2', 'NH3', 'NeH+', 'HCN', 'H2O', 'H7', 'H5', 'H8', 'C2H2', 'N2', 'Li2', 'H10', 'O3', 'N2H4', 'C2H6', 'N2H2', 'H2O2', 'CO2']
```

branch:
```pycon
>>> qml.data.load("qchem", molname="foo", basis="STO-3G", bondlength=0.86)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/data_manager/__init__.py", line 126, in load
    data_paths = [data_path for _, data_path in foldermap.find(data_name, **params)]
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/data_manager/foldermap.py", line 131, in find
    todo.extend(
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/data_manager/foldermap.py", line 134, in <genexpr>
    curr_level[next_param],
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/data/data_manager/foldermap.py", line 157, in __getitem__
    elem = self.__curr_level[__key]
KeyError: 'foo'
```

should be as simple as printing `list(self.__curr_level)` (except if `full` was provided for something earlier... but dw about that for now)

Reply:
^ yes, it is super helpful to print the available values from user's perspective.

Reply:
sure, added an error message for this